lib/src/rules/parameter_assignments.dart

```diff
@@ -124,75 +106,95 @@ class ParameterAssignments extends LintRule {
   }
 }
 
-class _Visitor extends SimpleAstVisitor<void> {
+class _DeclarationVisitor extends RecursiveAstVisitor {
+  final FormalParameter parameter;
   final LintRule rule;
+  final bool paramIsNotNullByDefault;
+  final bool paramDefaultsToNull;
+  bool hasBeenAssigned = false;
 
-  _Visitor(this.rule);
+  _DeclarationVisitor(this.parameter, this.rule)
+      : paramIsNotNullByDefault = parameter is SimpleFormalParameter ||
+            _isDefaultFormalParameterWithDefaultValue(parameter),
+        paramDefaultsToNull = parameter is DefaultFormalParameter &&
+            parameter.defaultValue == null;
```

Comment:
If both `paramIsNotNullByDefault` and `paramDefaultsToNull` are `false`, then there's no reason to visit the body of the function because we'll never report a lint. Consider computing these values outside the constructor and passing them in if we need to create the visitor. Doing so would avoid visiting bodies unnecessarily and creating unused objects, though the latter is presumably cheap because they'd be short-lived.

Reply:
Updated.  Thanks!